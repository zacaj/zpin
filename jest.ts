import { resetMachine, MachineOutputs, machine, Solenoid } from './machine';
import { resetSwitchMatrix } from './switch-matrix';
import { Events } from './events';
import { MPU } from './mpu';
import { Timer, setTime, wait } from './timer';
import { Log } from './log';

beforeEach(async () => {
    jest.spyOn(Timer, 'schedule').mockRestore();
    jest.spyOn(Events, 'listen').mockRestore();
    jest.spyOn(Events, 'fire').mockRestore();
    jest.spyOn(Timer, 'callIn').mockRestore();
    jest.spyOn(Log, 'init').mockImplementation(() => { throw 'unexpected' });
    jest.spyOn(Log, 'write').mockReturnValue();
    jest.spyOn(Log, 'logMessage').mockReturnValue();
    jest.spyOn(Log, 'trace').mockReturnValue();
    Timer.reset();
    Events.resetAll();
    resetSwitchMatrix();
    resetMachine();
    await setTime(1);
    jest.spyOn(MPU, 'sendCommandCode').mockImplementation(async (cmd) => {
        debugger;
        
        expect(cmd).toBe('mocked');

        return {
            code: 200,
            resp: 'mocked',
        };
    });
});

afterEach(async () => {
    Events.resetAll();
    Timer.reset();
    await setTime();
    jest.spyOn(Timer, 'callIn').mockReturnValue({} as any);
    jest.spyOn(Timer, 'schedule').mockReturnValue({} as any);
    jest.spyOn(Events, 'listen').mockReturnValue({} as any);
    jest.spyOn(Events, 'fire').mockReturnValue({} as any);
    await new Promise(r => setTimeout(r, 50));
});