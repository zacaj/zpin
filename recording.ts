import { Log } from './log';
import { Timer, passTime, setTime, time, wait } from './timer';
import * as fs from 'fs';
import { split, assert, debugging } from './util';
import { matrix } from './switch-matrix';
import { settleForks } from './promises';
import { MPU } from './mpu';
import { initMachine } from './init';
import { Events } from './events';
import { machine } from './machine';

export let curRecording: string|undefined = undefined;
let lines: string[];
let curTime = 0;
let curLine = 0;

export function initRecording(recording: string) {
    curRecording = recording;
    Log.log(['console'], 'playing back recording %s', recording);
    Timer.mockTime = 0;
    const file = fs.readFileSync(curRecording!, 'utf8');
    lines = file.split('\n');
    curTime = 0;
    curLine = 1;
}

export async function playRecording(toPoint?: string) {
    // await new Promise(r => setTimeout(r, 10000));
    const events = Events;
    const timer = Timer;
    const _machine = machine;
    for (; curLine < lines.length; curLine++) {
        const line = lines[curLine];
        if (line.startsWith('#')) continue;
        if (line && !line.includes(' ')) {
            const bp = line;
            if (bp === 'debug' && debugging()) {
                const nextLine = lines[curLine+1];
                debugger;
            } else if (process.env.NODE_ENV === 'test') {
                assert(bp === toPoint, `expected breakpoint ${toPoint} but got '${bp}'`);
                curLine++;
                await settleForks();
                return;
            }
            continue;
        }
        const [ts, sw, stateS, colS, rowS, timeS] = split(line, ' switch \'', '\' state -> ', ' (', ',', ') at ');
        if (timeS === undefined) continue;

        const time = parseInt(timeS, 10);
        const diff = time - curTime;
        assert(diff>=0, 'time travel not implemented');
        if (diff)
            await passTime(diff);
        curTime = time;

        matrix[parseInt(colS, 10)][parseInt(rowS, 10)]!.changeState(stateS === 'true', 'recording '+curRecording);
        await settleForks();
    }
    await settleForks();

    if (process.env.NODE_ENV !== 'test')
        await setTime(undefined);
    Log.log(['console', 'switch'], 'recording %s finished at %i', curRecording, time());
    curRecording = undefined;
}


export async function testRecording(name: string, toPoint?: string) {
    jest.spyOn(Log, 'init').mockImplementation(() => {});
    jest.spyOn(MPU, 'sendCommandCode').mockImplementation(async (cmd) => {
        return {
            code: 200,
            resp: 'mocked',
        };
    });
    await initMachine(false, false, true, false, `./recordings/${name}.rec`, toPoint);
}

export async function continueRecording(toPoint?: string) {
    await playRecording(toPoint);
}
export async function finishRecording() {
    await playRecording();
}
