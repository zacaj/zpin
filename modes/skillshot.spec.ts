import { machine } from '../machine';
import { testRecording, finishRecording } from '../recording';
import { snapshotOutputs, snapshotState, snapshot } from '../jest';

describe('skillshot', () => {
    test('close diverter on first switch hit', async () => {
        await testRecording('lower skillshot', 'break');
        expect(machine.cShooterDiverter.val).toBe(false);
        await finishRecording();
        expect(machine.cShooterDiverter.val).toBe(true);
        snapshot();
    });
});