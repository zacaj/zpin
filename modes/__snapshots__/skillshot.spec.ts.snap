// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`skillshot close diverter on first switch hit 1`] = `
Object {
  "outputs": Object {
    "centerBank": false,
    "getSkillshot": [Function],
    "iCenter1": "AD",
    "iCenter2": "KC",
    "iCenter3": "KS",
    "iLeft1": "8C",
    "iLeft2": "10S",
    "iLeft3": "JH",
    "iLeft4": "8S",
    "iMini1": "AH",
    "iMini2": "9S",
    "iMini3": "AC",
    "iRight1": "9C",
    "iRight2": "AS",
    "iRight3": "KH",
    "iRight4": "3D",
    "iRight5": "KD",
    "iSS1": "",
    "iSS2": "",
    "iSS3": "",
    "iSS4": "",
    "iSS5": "",
    "iSS6": "",
    "iSS7": "",
    "iSpinner": "",
    "iUpper21": "10H",
    "iUpper22": "QD",
    "iUpper31": "8D",
    "iUpper32": "10D",
    "iUpper33": "QS",
    "ignoreSkillsot": Set {},
    "kickerEnable": true,
    "lEjectArrow": Array [],
    "lEjectShowCards": Array [],
    "lEjectStartMode": Array [],
    "lLaneUpper1": Array [
      "Orange",
    ],
    "lLaneUpper2": Array [],
    "lLaneUpper3": Array [],
    "lLaneUpper4": Array [],
    "lLeftArrow": Array [],
    "lLowerRamp": Array [],
    "lMagnaSaveStatus": Array [
      "Green",
    ],
    "lMainTargetArrow": Array [],
    "lMiniReady": Array [
      "Red",
    ],
    "lPopperStatus": Array [
      "Green",
    ],
    "lPower1": Array [
      "Orange",
    ],
    "lPower2": Array [],
    "lPower3": Array [],
    "lPower4": Array [],
    "lRampArrow": Array [],
    "lRampShowCards": Array [],
    "lRampStartMb": Array [],
    "lShooterLaneArrow": Array [],
    "lShooterShowCards": Array [],
    "lShooterStartHand": Array [],
    "lSideShotArrow": Array [],
    "lSideTargetArrow": Array [],
    "lSpinnerArrow": Array [],
    "lUpperLaneArrow": Array [],
    "lUpperTargetArrow": Array [],
    "leftBank": false,
    "leftGate": false,
    "leftMagnet": false,
    "lockPost": false,
    "magnetPost": false,
    "miniBank": false,
    "miniDiverter": false,
    "miniEject": false,
    "miniFlipperEnable": false,
    "outhole": false,
    "popper": false,
    "rampUp": false,
    "realRightBank": false,
    "right1": false,
    "right2": false,
    "right3": false,
    "right4": false,
    "right5": false,
    "rightBank": false,
    "rightGate": true,
    "shooterDiverter": true,
    "troughRelease": false,
    "upper2": false,
    "upper3": false,
    "upperEject": false,
    "upperMagnet": false,
  },
  "state": Object {
    "children": Array [
      Object {
        "children": Array [],
        "name": "DropBank11",
        "state": Object {},
      },
      Object {
        "children": Array [],
        "name": "DropBank12",
        "state": Object {},
      },
      Object {
        "children": Array [],
        "name": "DropBank13",
        "state": Object {},
      },
      Object {
        "children": Array [],
        "name": "DropBank14",
        "state": Object {},
      },
      Object {
        "children": Array [],
        "name": "DropBank15",
        "state": Object {},
      },
      Object {
        "children": Array [],
        "name": "DropBank16",
        "state": Object {},
      },
      Object {
        "children": Array [
          Object {
            "children": Array [],
            "name": "EosPulse18",
            "state": Object {},
          },
        ],
        "name": "MachineOverrides17",
        "state": Object {},
      },
      Object {
        "children": Array [
          Object {
            "children": Array [
              Object {
                "children": Array [],
                "name": "ClearHoles21",
                "state": Object {},
              },
              Object {
                "children": Array [],
                "name": "Poker22",
                "state": Object {
                  "bet": 100,
                  "closeShooter": false,
                  "dealerCardsUsed": Array [],
                  "dealerHand": Array [
                    Object {
                      "flipped": true,
                      "num": 3,
                      "suit": "h",
                    },
                    Object {
                      "flipped": true,
                      "num": 4,
                      "suit": "c",
                    },
                    null,
                    null,
                    null,
                    null,
                    null,
                  ],
                  "dealerHandDesc": undefined,
                  "newMbs": Array [],
                  "newModes": Array [],
                  "playerCardsUsed": Array [],
                  "playerHand": Array [
                    Object {
                      "num": 7,
                      "suit": "s",
                    },
                    Object {
                      "num": 11,
                      "suit": "s",
                    },
                    null,
                    null,
                    null,
                    null,
                    null,
                  ],
                  "playerHandDesc": undefined,
                  "playerWins": undefined,
                  "pot": 400,
                  "slots": Array [
                    Object {
                      "num": 1,
                      "suit": "d",
                    },
                    Object {
                      "num": 13,
                      "suit": "c",
                    },
                    Object {
                      "num": 13,
                      "suit": "s",
                    },
                    Object {
                      "num": 9,
                      "suit": "c",
                    },
                    Object {
                      "num": 1,
                      "suit": "s",
                    },
                    Object {
                      "num": 13,
                      "suit": "h",
                    },
                    Object {
                      "num": 3,
                      "suit": "d",
                    },
                    Object {
                      "num": 13,
                      "suit": "d",
                    },
                    Object {
                      "num": 8,
                      "suit": "c",
                    },
                    Object {
                      "num": 10,
                      "suit": "s",
                    },
                    Object {
                      "num": 11,
                      "suit": "h",
                    },
                    Object {
                      "num": 8,
                      "suit": "s",
                    },
                    Object {
                      "num": 10,
                      "suit": "h",
                    },
                    Object {
                      "num": 12,
                      "suit": "d",
                    },
                    Object {
                      "num": 8,
                      "suit": "d",
                    },
                    Object {
                      "num": 10,
                      "suit": "d",
                    },
                    Object {
                      "num": 12,
                      "suit": "s",
                    },
                    Object {
                      "num": 1,
                      "suit": "h",
                    },
                    Object {
                      "num": 9,
                      "suit": "s",
                    },
                    Object {
                      "num": 1,
                      "suit": "c",
                    },
                  ],
                  "step": 2,
                },
              },
              Object {
                "children": Array [],
                "name": "Spinner23",
                "state": Object {
                  "comboMult": 1,
                  "rounds": 0,
                  "score": 10,
                },
              },
              Object {
                "children": Array [],
                "name": "LeftOrbit24",
                "state": Object {
                  "comboMult": 1,
                  "rounds": 1,
                  "score": 20000,
                },
              },
              Object {
                "children": Array [],
                "name": "PlayerOverrides25",
                "state": Object {},
              },
              Object {
                "children": Array [
                  Object {
                    "children": Array [
                      Object {
                        "children": Array [],
                        "name": "DropBankResetter28",
                        "state": Object {},
                      },
                      Object {
                        "children": Array [],
                        "name": "DropBankResetter29",
                        "state": Object {},
                      },
                      Object {
                        "children": Array [],
                        "name": "DropBankResetter30",
                        "state": Object {},
                      },
                      Object {
                        "children": Array [],
                        "name": "DropBankResetter31",
                        "state": Object {},
                      },
                      Object {
                        "children": Array [],
                        "name": "DropBankResetter32",
                        "state": Object {},
                      },
                      Object {
                        "children": Array [],
                        "name": "DropBankResetter33",
                        "state": Object {},
                      },
                    ],
                    "name": "ResetAnyDropOnComplete27",
                    "state": Object {},
                  },
                ],
                "name": "Ball26",
                "state": Object {
                  "miniReady": false,
                },
              },
            ],
            "name": "Player20",
            "state": Object {
              "chips": 1,
              "closeShooter": false,
              "curMbMode": undefined,
              "laneChips": Array [
                true,
                false,
                false,
                false,
              ],
              "mbsQualified": Array [],
              "modesQualified": Array [],
              "noMode": undefined,
              "poker": "Poker22",
              "rampUp": false,
              "score": 11010,
            },
          },
        ],
        "name": "Game19",
        "state": Object {
          "ballNum": 1,
        },
      },
    ],
    "name": "Machine10",
    "state": Object {
      "lastSwitchHit": Switch {
        "_state": false,
        "column": 5,
        "lastChange": 10367,
        "lastClosed": 10077,
        "lastOpened": 10367,
        "minOffTime": 10,
        "minOnTime": 1,
        "name": "right inlane",
        "row": 0,
      },
      "lockDown": false,
      "miniDown": false,
    },
  },
}
`;
